File: 프로그래머스/3/258707. n ＋ 1 카드게임/n ＋ 1 카드게임.py
from collections import deque

def make_target(cards1, cards2, target):
    for card in cards1:
        if target - card in cards2:
            cards1.remove(card)
            cards2.remove(target - card)
            return True
    return False

def solution(coin, cards):
    answer = 1
    hand = set(cards[:len(cards) // 3])
    left_cards = deque(cards[len(cards) // 3:])
    passed = set()
    target = len(cards) + 1
    
    while left_cards:
        passed.add(left_cards.popleft())
        passed.add(left_cards.popleft())
        
        # 손에 있는 카드 두 장으로 target 만들기
        if make_target(hand, hand, target):
            answer += 1
            continue
        
        # 손에 있는 카드 1장 + 뽑았던 카드 중에 1장 (coin -= 1)
        elif coin >= 1 and make_target(hand, passed, target):
            coin -= 1
            answer += 1
            continue
        
        # 뽑았던 카드 중 2장 (coin -= 2)
        elif coin >= 2 and make_target(passed, passed, target):
            coin -= 2
            answer += 1
            continue
        
        # 위 모든 경우가 불가능하면 종료
        else:
            break
            
    return answer
    

File: 프로그래머스/3/258707. n ＋ 1 카드게임/n ＋ 1 카드게임.java
import java.util.*;

class Solution {
    public int solution(int coin, int[] cards) {
        int n = cards.length;
        
        var hand = new HashSet<Integer>();
        var passed = new HashSet<Integer>();
        var cardQ = new ArrayDeque<Integer>();
        
        for (int card: cards) cardQ.addLast(card);
        for (int i = 0; i < n / 3; i++) hand.add(cardQ.removeFirst());
        
        int round = 1;
        while (cardQ.size() >= 2) {
            passed.add(cardQ.removeFirst());
            passed.add(cardQ.removeFirst());
            
            boolean goToNextRound = false;
            
            // 손에 있는 카드 두 장을 내고 다음 라운드 진출 (coin 소모 X)
            for (int cardInHand: hand) {
                if (hand.contains(n + 1 - cardInHand)) {
                    hand.remove(cardInHand);
                    hand.remove(n + 1 - cardInHand);
                    goToNextRound = true;
                    break;
                }
            }
            if (goToNextRound) {
                round += 1;
                continue;
            }
            
            // 손에 있는 카드 + 지나간 카드 한 장을 내고 다음 라운드 진출 (coin -= 1)
            for (int cardInHand: hand) {
                if (passed.contains(n + 1 - cardInHand) && coin >= 1) {
                    hand.remove(cardInHand);
                    passed.remove(n + 1 - cardInHand);
                    coin -= 1;
                    goToNextRound = true;
                    break;
                }
            }
            if (goToNextRound) {
                round += 1;
                continue;
            }
            
            // 지나간 카드 두 장을 내고 다음 라운드 진출 (coin -= 2)
            for (int cardInPassed: passed) {
                if (passed.contains(n + 1 - cardInPassed) && coin >= 2) {
                    passed.remove(cardInPassed);
                    passed.remove(n + 1 - cardInPassed);
                    coin -= 2;
                    goToNextRound = true;
                    break;
                }
            }
            if (goToNextRound) {
                round += 1;
                continue;
            }
            
            // 전부 불가능하면 게임 종료
            break;
        }
        
        return round;
    }
}